### ISSUE TITLE
parallel-workload: thread 'coordinator' panicked at src/adapter/src/catalog/apply.rs:825:30: invalid persisted SQL: {create_sql}: PlanError(Catalog(UnknownSchema("s-3-49")))

### ISSUE THREAD

**Issue by def-** [MaterializeInc/materialize#28640]
### What version of Materialize are you using?

6fcb05b2b0

### What is the issue?

Seen in [Parallel Workload (rename)](https://buildkite.com/materialize/nightly/builds/8866#01910369-157c-42a1-899b-cb1f22102b5f), probably because the test runs further now and always deadlocked before, so this might not be a product regression:
```
thread 'coordinator' panicked at src/adapter/src/catalog/apply.rs:825:30:
invalid persisted SQL: {create_sql}: PlanError(Catalog(UnknownSchema("s-3-49")))
stack backtrace:
   0: rust_begin_unwind
             at ./rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/panicking.rs:652:5
   1: core::panicking::panic_fmt
             at ./rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/panicking.rs:72:14
   2: core::result::unwrap_failed
             at ./rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/result.rs:1654:5
   3: <core::result::Result<mz_catalog::memory::objects::CatalogItem, mz_adapter::error::AdapterError>>::expect
             at ./rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/result.rs:1034:23
   4: <mz_adapter::catalog::state::CatalogState>::apply_item_update
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-00d97154d6ec58e92-1/materialize/test/src/adapter/src/catalog/apply.rs:823:44
   5: <mz_adapter::catalog::state::CatalogState>::apply_update
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-00d97154d6ec58e92-1/materialize/test/src/adapter/src/catalog/apply.rs:228:17
   6: <mz_adapter::catalog::state::CatalogState>::apply_updates_inner
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-00d97154d6ec58e92-1/materialize/test/src/adapter/src/catalog/apply.rs:171:21
   7: <mz_adapter::catalog::state::CatalogState>::apply_updates
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-00d97154d6ec58e92-1/materialize/test/src/adapter/src/catalog/apply.rs:141:17
   8: <mz_adapter::catalog::Catalog>::transact_inner::{closure#0}::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-00d97154d6ec58e92-1/materialize/test/src/adapter/src/catalog/transact.rs:511:44
   9: <tracing::instrument::Instrumented<<mz_adapter::catalog::Catalog>::transact_inner::{closure#0}::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  10: <mz_adapter::catalog::Catalog>::transact_inner::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-00d97154d6ec58e92-1/materialize/test/src/adapter/src/catalog/transact.rs:443:5
  11: <mz_adapter::catalog::Catalog>::transact::{closure#0}::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-00d97154d6ec58e92-1/materialize/test/src/adapter/src/catalog/transact.rs:404:10
  12: <tracing::instrument::Instrumented<<mz_adapter::catalog::Catalog>::transact::{closure#0}::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  13: <mz_adapter::catalog::Catalog>::transact::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-00d97154d6ec58e92-1/materialize/test/src/adapter/src/catalog/transact.rs:340:5
  14: <mz_adapter::coord::Coordinator>::catalog_transact_inner::{closure#0}::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-00d97154d6ec58e92-1/materialize/test/src/adapter/src/coord/ddl.rs:517:14
  15: <tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::catalog_transact_inner::{closure#0}::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  16: <mz_adapter::coord::Coordinator>::catalog_transact_inner::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-00d97154d6ec58e92-1/materialize/test/src/adapter/src/coord/ddl.rs:182:5
  17: <mz_adapter::coord::Coordinator>::catalog_transact_with_side_effects::<<mz_adapter::coord::Coordinator>::create_index_finish::{closure#0}::{closure#0}::{closure#1}, <mz_adapter::coord::Coordinator>::create_index_finish::{closure#0}::{closure#0}::{closure#1}::{closure#0}>::{closure#0}::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-00d97154d6ec58e92-1/materialize/test/src/adapter/src/coord/ddl.rs:96:14
  18: <tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::catalog_transact_with_side_effects<<mz_adapter::coord::Coordinator>::create_index_finish::{closure#0}::{closure#0}::{closure#1}, <mz_adapter::coord::Coordinator>::create_index_finish::{closure#0}::{closure#0}::{closure#1}::{closure#0}>::{closure#0}::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  19: <mz_adapter::coord::Coordinator>::catalog_transact_with_side_effects::<<mz_adapter::coord::Coordinator>::create_index_finish::{closure#0}::{closure#0}::{closure#1}, <mz_adapter::coord::Coordinator>::create_index_finish::{closure#0}::{closure#0}::{closure#1}::{closure#0}>::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-00d97154d6ec58e92-1/materialize/test/src/adapter/src/coord/ddl.rs:83:5
  20: <mz_adapter::coord::Coordinator>::create_index_finish::{closure#0}::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-00d97154d6ec58e92-1/materialize/test/src/adapter/src/coord/sequencer/inner/create_index.rs:516:14
  21: <tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::create_index_finish::{closure#0}::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  22: <mz_adapter::coord::Coordinator>::create_index_finish::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-00d97154d6ec58e92-1/materialize/test/src/adapter/src/coord/sequencer/inner/create_index.rs:399:5
  23: <mz_adapter::coord::CreateIndexStage as mz_adapter::coord::Staged>::stage::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-00d97154d6ec58e92-1/materialize/test/src/adapter/src/coord/sequencer/inner/create_index.rs:57:65
  24: <tracing::instrument::Instrumented<<mz_adapter::coord::CreateIndexStage as mz_adapter::coord::Staged>::stage::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  25: <mz_adapter::coord::Coordinator>::sequence_staged::<mz_adapter::coord::CreateIndexStage>::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-00d97154d6ec58e92-1/materialize/test/src/adapter/src/coord/sequencer/inner.rs:199:18
  26: <mz_adapter::coord::Coordinator>::handle_message::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-00d97154d6ec58e92-1/materialize/test/src/adapter/src/coord/message_handler.rs:161:60
  27: <tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::handle_message::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  28: <core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>> as core::future::future::Future>::poll
             at ./rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/future/future.rs:123:9
  29: <mz_adapter::coord::Coordinator>::serve::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-00d97154d6ec58e92-1/materialize/test/src/adapter/src/coord.rs:2975:48
  30: <core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>> as core::future::future::Future>::poll
             at ./rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/future/future.rs:123:9
  31: <tracing::instrument::Instrumented<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  32: <tokio::runtime::park::CachedParkThread>::block_on::<tracing::instrument::Instrumented<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>>>::{closure#0}
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/park.rs:281:63
  33: tokio::runtime::coop::with_budget::<core::task::poll::Poll<()>, <tokio::runtime::park::CachedParkThread>::block_on<tracing::instrument::Instrumented<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>>>::{closure#0}>
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/coop.rs:107:5
  34: tokio::runtime::coop::budget::<core::task::poll::Poll<()>, <tokio::runtime::park::CachedParkThread>::block_on<tracing::instrument::Instrumented<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>>>::{closure#0}>
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/coop.rs:73:5
  35: <tokio::runtime::park::CachedParkThread>::block_on::<tracing::instrument::Instrumented<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>>>
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/park.rs:281:31
  36: <tokio::runtime::context::blocking::BlockingRegionGuard>::block_on::<tracing::instrument::Instrumented<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>>>
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/context/blocking.rs:66:9
  37: <tokio::runtime::handle::Handle>::block_on::<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>>::{closure#0}
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/handle.rs:310:22
  38: tokio::runtime::context::runtime::enter_runtime::<<tokio::runtime::handle::Handle>::block_on<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>>::{closure#0}, ()>
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/context/runtime.rs:65:16
  39: <tokio::runtime::handle::Handle>::block_on::<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()>>>>
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.0/src/runtime/handle.rs:309:9
  40: mz_adapter::coord::serve::{closure#0}::{closure#2}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-00d97154d6ec58e92-1/materialize/test/src/adapter/src/coord.rs:3606:21
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```
```
ci-regexp: invalid persisted SQL: 
```

**Comment by jkosh44** [comment]
Just from looking at the error message, this looks identical to the previous issue we had. It looked a little something like this:

1. Start DDL (`A`) that references schema `s` in the create SQL statement.
2. `A` does some work off-thread.
3. Someone else executes `ALTER SCHEMA s RENAME TO _`.
4. `A` completes off thread work and is committed to the catalog with a reference to schema `s` which no longer exists.

I have no actual proof that this is what's happening, but it's a good place to start.

**Comment by jkosh44** [comment]
This PR should at least fix the error message so we can look at the invalid SQL: https://github.com/MaterializeInc/materialize/pull/28645

**Comment by jkosh44** [comment]
Two extremely suspicious log lines right before the panic are

```
parallel-workload-materialized-1     | environmentd: 2024-07-30T12:02:02.574164Z  INFO coord::handle_message{kind="command-execute"}:message_command:coord::handle_execute{session="794dedc9-cc38-45cb-b27d-c72c214d0c40"}:coord::handle_execute_inner{stmt="ALTER SCHEMA \"db-pw-1722339381-1\".\"s-3-49\" RENAME TO \"s-3-50\""}:sequence_alter_schema_rename:coord::catalog_transact_with_ddl_transaction:coord::catalog_transact:coord::catalog_transact_conn:coord::catalog_transact_inner:catalog::transact:catalog::transact_inner:transact_op: 

parallel-workload-materialized-1     | environmentd: 2024-07-30T12:02:03.001028Z  INFO coord::handle_message{kind="deferred_statement_ready"}:handle_deferred_statement:coord::handle_execute_inner{stmt="CREATE INDEX idx_kafka_table4_8931525370510405289 ON \"db-pw-1722339381-1\".\"s-3-49\".kafka_table4 (value3, value2, key1, value1, key5, key2, key0, key3, key4, value0)"}:sequence_create_index:create_index_finish:coord::catalog_transact_with_side_effects:coord::catalog_transact_inner:catalog::transact:catalog::transact_inner:transact_op: mz_adapter::catalog::transact: create index db-pw-1722339381-1.s-3-50.idx_kafka_table4_8931525370510405289 (u417)
```

The `CREATE INDEX` log line contains both the old and the new schema name.

**Comment by jkosh44** [comment]
This same repo works for me to reproduce the issue on `main`.

**Comment by jkosh44** [comment]
The way that the DDL lock works is:

1. Parse DDL.
2. Try and acquire DDL lock.
3. If you can't acquire lock, then add deferred DDL statement to a queue.
4. Once the lock is available, then pop the statement and execute it.

The deferred DDL statement struct looks like this:

https://github.com/MaterializeInc/materialize/blob/82c077dc91e368ed11470ea679f701b97806be27/src/adapter/src/coord.rs#L3975-L3991

From the logs it looks like the deferred statement already has the old schema name in it. So after it's popped from the queue the create SQL is invalid but somehow it still successfully makes it through execution. Then it commits the invalid SQL to the catalog.

**Comment by jkosh44** [comment]
The plan variant of the `DeferredPlanStatement` contains CREATE SQL statements.

For example here's a deferred `CREATE INDEX`.

https://github.com/MaterializeInc/materialize/blob/14482dd5c7249db04e8c31fde7cf1b8bbbe4b91d/src/adapter/src/coord.rs#L3996-L4012

https://github.com/MaterializeInc/materialize/blob/14482dd5c7249db04e8c31fde7cf1b8bbbe4b91d/src/sql/src/plan.rs#L112-L194

https://github.com/MaterializeInc/materialize/blob/14482dd5c7249db04e8c31fde7cf1b8bbbe4b91d/src/sql/src/plan.rs#L696-L701

https://github.com/MaterializeInc/materialize/blob/14482dd5c7249db04e8c31fde7cf1b8bbbe4b91d/src/sql/src/plan.rs#L1495-L1502

I can think of three potential solutions.

1. Regenerate the SQL in a deferred plan if the plan is still valid. While nice in theory, I'm not sure that it's possible.
2. Validate the SQL by re-planning and parsing it and returning an error if it fails. I don't like this solution because we end up rejecting potentially valid plans if objects are renamed, and it requires both OCC and pessimistic concurrency control.
3. Always take the DDL lock before name resolution. I think that this is the best approach.

**Comment by jkosh44** [comment]
Ok, turns out my assessment was completely wrong. The issue is that renames never acquire the DDL lock. The reason is this line: https://github.com/MaterializeInc/materialize/blob/ea4d9adfeea49e99da74720d9d322b5480c1241c/src/adapter/src/coord/command_handler.rs#L981-L984

I'm not really sure what the meaning of that comment is, but it doesn't seem to be true.

**Comment by jkosh44** [comment]
The issue is that when a rename is executed in an explicit transaction then we add `TransactionOps::DDL` to the current transaction. That causes the following check to return `true`:

https://github.com/MaterializeInc/materialize/blob/ea4d9adfeea49e99da74720d9d322b5480c1241c/src/adapter/src/coord/sequencer.rs#L282-L309

https://github.com/MaterializeInc/materialize/blob/ea4d9adfeea49e99da74720d9d322b5480c1241c/src/adapter/src/session.rs#L999-L1010

which causes us to try and acquire the DDL lock on commit. If we had also tried to get the DDL lock when executing the rename plan, then we'd end up deadlocking.

So basically we acquire the DDL lock for renames if that rename is in an explicit transaction, but not in an implicit transaction. We need to devise a way to acquire the lock exactly once for both situations.
